[16.0] auth_jwt: add cookie mode

sbidoul opened the pull request:
Add a cookie mode on JWT validators. In this case, the JWT payload obtained
from the ``Authorization`` header is returned as a Http-Only cookie. This mode is
sometimes simpler for front-end applications which do not then need to store and protect
the JWT token across requests and can simply rely on the cookie management mechanisms of
browsers. When both the ``Authorization`` header and a cookie are provided, the cookie
is ignored in order to let clients authenticate with a different user by providing a new
JWT token.

sbidoul committed:
auth_jwt: add cookie mode

sbidoul commented:
I keep this draft until the design of `auth_jwt_fastapi` in oca/rest-framework is ready, as we may need some refactoring.

hparfr reviewed (approved):
code read, didn't test

sbidoul committed:
auth_jwt: clarify exceptions

Distinguish errors that lead to a 401
from internal configuration errors.

sbidoul committed:
auth_jwt: minor refactoring

sbidoul committed:
[IMP] auth_jwt: refactor

Extract _parse_bearer_authorization function for easier reuse by fastapi_auth_jwt

sbidoul committed:
[FIX] auth_jwt: don't use public mode if a cookie is present

sbidoul committed:
[IMP] auth_jwt: split test class

sbidoul commented:
This should now be ready.

sbidoul committed:
[IMP] auth_jwt: check cookie_name is present in cookie mode

lmignon commented on auth_jwt/models/ir_http.py:
```diff
@@ -54,12 +52,25 @@ def _authenticate(cls, endpoint):
+        cookie."""
+        try:
+            token = cls._get_bearer_token()
+            assert token
```
```suggestion
            assert token, "JWT token not found into request header nor as cookie"  
```

lmignon reviewed (approved):
Great. LGTM (Code review)

sbidoul commented:
Thanks for the review, @lmignon @hparfr. 

/ocabot merge minor
